micro-rdk/src/common/data_manager.rs

```diff
@@ -0,0 +1,402 @@
+    sync_interval_ms: u64,
+    part_id: String,
+}
+
+impl<StoreType> DataManager<StoreType>
+where
+    StoreType: DataStore,
+{
+    pub fn new(
+        collectors: Vec<DataCollector>,
+        store: StoreType,
+        sync_interval_ms: u64,
+        part_id: String,
+    ) -> Self {
+        Self {
+            collectors,
+            store,
+            sync_interval_ms,
+            part_id,
+        }
+    }
+
+    pub fn sync_interval_ms(&self) -> u64 {
+        self.sync_interval_ms
+    }
+
+    pub fn part_id(&self) -> String {
+        self.part_id.clone()
+    }
+
+    pub(crate) fn collection_intervals(&self) -> Vec<u64> {
+        let mut intervals: Vec<u64> = self.collectors.iter().map(|x| x.time_interval()).collect();
+        intervals.sort();
+        intervals.dedup();
+        intervals
+    }
+
+    pub async fn run(&mut self) -> Result<(), DataManagerError> {
+        let min_interval = self
+            .collection_intervals()
+            .into_iter()
+            .min()
+            .ok_or(DataManagerError::NoCollectors)?;
+        let min_interval = std::cmp::min(min_interval, self.sync_interval_ms);
+        let mut loop_counter: u64 = 0;
+        loop {
+            Timer::after(Duration::from_millis(min_interval)).await;
+            if (self.sync_interval_ms / min_interval) % loop_counter == 0 {
+                self.sync()?;
+            }
+            for interval in self.collection_intervals() {
+                if (interval / min_interval) % loop_counter == 0 {
+                    self.collect_and_store_readings(interval)?;
+                }
+            }
+            loop_counter += 1;
+        }
+    }
+
+    fn sync(&mut self) -> Result<(), DataManagerError> {
+        for collector_key in self.collectors.iter().map(|c| c.resource_method_key()) {
+            // TODO: check for internet access before attempting to read from store
+            let mut readings_to_upload: Vec<BytesMut> = vec![];
+            loop {
+                match self.store.read_next_message(&collector_key) {
+                    Ok(msg) => {
+                        if msg.is_empty() {
+                            break;
+                        } else {
+                            readings_to_upload.push(msg);
+                        }
+                    }
+                    Err(err) => return Err(err.into()),
+                };
+            }
+            // TODO: implement actual upload logic here, will likely have to change struct
+            // and make this function async
+        }
+        Ok(())
+    }
+
+    pub fn collect_and_store_readings(
+        &mut self,
+        time_interval_ms: u64,
+    ) -> Result<(), DataManagerError> {
+        for (collector_key, reading) in self.collect_readings_for_interval(time_interval_ms)? {
+            self.store
+                .write_message(&collector_key, reading, WriteMode::OverwriteOldest)?
+        }
+        Ok(())
+    }
+
+    fn collect_readings_for_interval(
```

Comment:
So we will end up doing #unique-collection-intervals scans of the collectors array? So, potentially N^2 in number of collectors if all the intervals are different? To be clear, I'm not convinced that is really a problem since I expect N to be very small. Just noting it.

Reply:
Yes, I've taken to programming less computationally efficient solutions for micro-rdk when I believe there's memory to be saved because the number of elements we deal with is usually small. But I may be overlooking something here, I'll double-check the logic

Reply:
Couldn't think of a way to reduce time complexity without some sort of caching (which wouldn't seem worth it)